Source/Core/Core/State.h

```diff
@@ -20,15 +21,40 @@ static const u32 NUM_STATES = 10;
 struct StateHeader
 {
   char gameID[6];
-  u16 reserved1;
-  u32 size;
-  u32 reserved2;
+  char reserved1[2];
+  u32 must_be_zero = 0;  // prevent decompression so older Dolphin can read state version info
+  char reserved2[4];
   double time;
+  u32 version_cookie;
+  u32 version_str_len;
+  std::unique_ptr<char[]> version_str;
+
+  int FileLength() { return offsetof(StateHeader, version_str) + version_str_len; }
```

Comment:
I suspect it is not legal in C++ to have a partially-copyable types like this. You should probably split this up into multiple nested structs.